tests/unit/s2n_rsa_pss_rsae_test.c

```diff
@@ -245,11 +247,20 @@ int main(int argc, char **argv)
             EXPECT_SUCCESS(s2n_asn1der_to_public_key_and_type(&rsa_pss_public_key, &rsa_pss_pkey_type, &rsa_pss_cert_chain->cert_chain->head->raw));
             EXPECT_EQUAL(rsa_pss_pkey_type, S2N_PKEY_TYPE_RSA_PSS);
 
-            /* Set the keys equal */
+            /* Set the keys equal.
+             *
+             * In openssl 3 the function `EVP_PKEY_get1_RSA` returns a pre-cached copy of PKEY
+             * rather than the original copy. For this reason it is not possible to modify the
+             * original underlying RSA key.
+             *
+             * As a workaround, `rsa_key_copy` is modified, `rsa_pss_public_key` is set using
+             * `EVP_PKEY_set1_RSA` and finally rsa_key_copy is freed. This change is also
+             * backwards compatible with other ossl impl down to 1.0.2. */
+            RSA *rsa_key_copy = EVP_PKEY_get1_RSA(rsa_public_key.pkey);
             const BIGNUM *n, *e, *d;
-            RSA_get0_key(EVP_PKEY_get0_RSA(rsa_public_key.pkey), &n, &e, &d);
-            EXPECT_SUCCESS(RSA_set0_key(EVP_PKEY_get0_RSA(rsa_pss_public_key.pkey),
-                    BN_dup(n), BN_dup(e), BN_dup(d)));
+            RSA_get0_key(rsa_key_copy, &n, &e, &d);
+            POSIX_GUARD_OSSL(EVP_PKEY_set1_RSA(rsa_pss_public_key.pkey, rsa_key_copy), S2N_ERR_KEY_INIT);
+            RSA_free(rsa_key_copy);
```

Comment:
Okay I actually feel like I finally understand what's going on with the above code. We're retrieving the rsa key, storing its bignum values, and using those values to set the rsa_pss key values. 🙄 Let's never daisy-chain Openssl functions again.

Anyways something about this change doesn't look right to me. We are setting n, e, and d on line 261 and then never using them again. Can you just remove line 260 and 261?

Reply:
Yep, seems like we dont use those anymore. Good catch.

I think some of this code goes back pretty far so its good to finally bring this more up to date. rsa_pss is the only place were we use `EVP_PKEY_get0_...` instead of `EVP_PKEY_get1_...`.